File: sdks/python/tyk_fapi/jwks_server.py
"""
JWKS Server implementation for the Tyk FAPI SDK.
"""

import json
import logging
import threading
from flask import Flask, jsonify, request
from typing import Dict, List, Optional, Callable

from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePublicKey
from .utils import public_key_to_jwk


class JWKSServer:
    """
    A server that hosts a JSON Web Key Set (JWKS) endpoint.
    """
    
    def __init__(self, public_key: EllipticCurvePublicKey, kid: str, port: int = 8082):
        """
        Initialize the JWKS server.
        
        Args:
            public_key: The public key to serve.
            kid: The key ID.
            port: The port to run the server on.
        """
        self.public_key = public_key
        self.kid = kid
        self.port = port
        self.app = Flask(__name__)
        self.server_thread = None
        self.jwks_url = f"http://localhost:{port}/.well-known/jwks.json"
        
        # Set up routes
        self.app.route("/.well-known/jwks.json")(self.handle_jwks)
        
        # Disable Flask's default logging
        log = logging.getLogger('werkzeug')
        log.setLevel(logging.ERROR)
    
    def handle_jwks(self):
        """
        Handle requests to the JWKS endpoint.
        
        Returns:
            A JSON response containing the JWKS.
        """
        # Log the request
        print(f"JWKS request from {request.remote_addr}")
        
        # Create a JWK from the public key
        jwk = public_key_to_jwk(self.public_key, self.kid)
        
        # Create a JWKS
        jwks = {
            "keys": [jwk]
        }
        
        # Set CORS headers
        response = jsonify(jwks)
        response.headers.add("Access-Control-Allow-Origin", "*")
        response.headers.add("Access-Control-Allow-Methods", "GET, OPTIONS")
        response.headers.add("Access-Control-Allow-Headers", "Content-Type")
        
        return response
    
    def start(self) -> None:
        """
        Start the JWKS server in a separate thread.
        
        Raises:
            RuntimeError: If the server is already running.
        """
        if self.server_thread is not None and self.server_thread.is_alive():
            raise RuntimeError("JWKS server already running")
        
        def run_server():
            self.app.run(host="0.0.0.0", port=self.port, debug=False, use_reloader=False)
        
        self.server_thread = threading.Thread(target=run_server, daemon=True)
        self.server_thread.start()
        
        print(f"Starting JWKS server on port {self.port}")
        print(f"JWKS server started. Local URL: {self.jwks_url}")
        print(f"For Keycloak in Docker, use: http://host.docker.internal:{self.port}/.well-known/jwks.json")
    
    def stop(self) -> None:
        """
        Stop the JWKS server.
        """
        if self.server_thread is None or not self.server_thread.is_alive():
            return
        
        # This is a bit of a hack, but Flask doesn't provide a clean way to stop the server
        import requests
        try:
            requests.get(f"http://localhost:{self.port}/shutdown", timeout=0.1)
        except requests.exceptions.RequestException:
            pass
        
        self.server_thread = None
        print("JWKS server stopped")

File: sdks/nodejs/src/jwksServer.ts
/**
 * JWKS Server implementation for the Tyk FAPI SDK.
 */

import * as express from 'express';
import * as http from 'http';
import * as jose from 'jose';
import { JWK, JWKS } from './types';
import { publicKeyToJWK } from './utils';

/**
 * A server that hosts a JSON Web Key Set (JWKS) endpoint.
 */
export class JWKSServer {
  /**
   * The public key to serve.
   */
  private publicKey: jose.KeyLike;
  
  /**
   * The key ID.
   */
  private kid: string;
  
  /**
   * The port to run the server on.
   */
  private port: number;
  
  /**
   * The Express app.
   */
  private app: express.Express;
  
  /**
   * The HTTP server.
   */
  private server: http.Server | null = null;
  
  /**
   * The JWKS URL.
   */
  public jwksUrl: string;
  
  /**
   * Initialize the JWKS server.
   * 
   * @param publicKey - The public key to serve.
   * @param kid - The key ID.
   * @param port - The port to run the server on.
   */
  constructor(publicKey: jose.KeyLike, kid: string, port: number = 8082) {
    this.publicKey = publicKey;
    this.kid = kid;
    this.port = port;
    this.app = express();
    this.jwksUrl = `http://localhost:${port}/.well-known/jwks.json`;
    
    // Set up routes
    this.app.get('/.well-known/jwks.json', this.handleJwks.bind(this));
  }
  
  /**
   * Handle requests to the JWKS endpoint.
   * 
   * @param req - The request object.
   * @param res - The response object.
   */
  private async handleJwks(req: express.Request, res: express.Response): Promise<void> {
    // Log the request
    console.log(`JWKS request from ${req.ip}`);
    
    try {
      // Create a JWK from the public key
      const jwk = await publicKeyToJWK(this.publicKey, this.kid);
      
      // Create a JWKS
      const jwks: JWKS = {
        keys: [jwk]
      };
      
      // Set CORS headers
      res.header('Access-Control-Allow-Origin', '*');
      res.header('Access-Control-Allow-Methods', 'GET, OPTIONS');
      res.header('Access-Control-Allow-Headers', 'Content-Type');
      
      // Send the response
      res.json(jwks);
    } catch (error) {
      console.error('Error handling JWKS request:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  }
  
  /**
   * Start the JWKS server.
   * 
   * @returns A Promise that resolves when the server is started.
   * @throws Error if the server is already running.
   */
  public start(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this.server) {
        reject(new Error('JWKS server already running'));
        return;
      }
      
      this.server = this.app.listen(this.port, () => {
        console.log(`Starting JWKS server on port ${this.port}`);
        console.log(`JWKS server started. Local URL: ${this.jwksUrl}`);
        console.log(`For Keycloak in Docker, use: http://host.docker.internal:${this.port}/.well-known/jwks.json`);
        resolve();
      });
      
      this.server.on('error', (error) => {
        this.server = null;
        reject(error);
      });
    });
  }
  
  /**
   * Stop the JWKS server.
   * 
   * @returns A Promise that resolves when the server is stopped.
   */
  public stop(): Promise<void> {
    return new Promise((resolve) => {
      if (!this.server) {
        resolve();
        return;
      }
      
      this.server.close(() => {
        this.server = null;
        console.log('JWKS server stopped');
        resolve();
      });
    });
  }
}